### PR TITLE
Support for auto-generated ObjectId

### PR DESCRIPTION
**Pull request by damieng**
Fixes #92

This PR includes some clean up and refactoring which has been separated out into individual commits for easy review.

### CODE REVIEW

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Update/AddEntityTests.cs [hunk 1]
```diff
@@ -70,26 +70,57 @@ class MongoSpecificTypeEntity
...
+    public void Add_simple_entity_with_unset_ObjectId()
+    {
+        var collection = _tempDatabase.CreateTemporaryCollection<SimpleEntity>();
+        var dbContext = SingleEntityDbContext.Create(collection);
+
+        var expected = new SimpleEntity {name = "Not Set"};
+        dbContext.Entitites.Add(expected);
+        dbContext.SaveChanges();
+
+        Assert.Same(expected, dbContext.Entitites.First());
```
Comment (adelinowona) [inline tests/MongoDB.EntityFrameworkCore.FunctionalTests/Update/AddEntityTests.cs:100]: Probably assert that the objectId is not all zeros here.

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Update/AddEntityTests.cs [hunk 2]
```diff
@@ -70,26 +70,57 @@ class MongoSpecificTypeEntity
...
+    public void Add_simple_entity_with_empty_ObjectId()
+    {
+        var collection = _tempDatabase.CreateTemporaryCollection<SimpleEntity>();
+        var dbContext = SingleEntityDbContext.Create(collection);
+
+        var expected = new SimpleEntity {_id = ObjectId.Empty, name = "Empty"};
+        dbContext.Entitites.Add(expected);
+        dbContext.SaveChanges();
+
+        Assert.Same(expected, dbContext.Entitites.First());
```
Comment (adelinowona) [inline tests/MongoDB.EntityFrameworkCore.FunctionalTests/Update/AddEntityTests.cs:118]: Same as previous comment, add an assert for objectId is not all zeros.

**Review by adelinowona — APPROVED** [review]
LGTM with minor changes.

File: src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
+﻿/* Copyright 2024-present MongoDB Inc.
```
Comment (sanych-sun) [inline src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs:1]: 2024 or 2023? in Driver we use 2010 everywhere even for new files.

File: src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    /// <param name="property">The property determine a generation strategy for.</param>
+    /// <returns>
+    /// The store <see cref="ValueGenerated"/> strategy to set for the <paramref name="property"/>
+    /// or <see langref="null"/> if no strategy should be used.
+    /// </returns>
+    protected override ValueGenerated? GetValueGenerated(IConventionProperty property)
+    {
+        if (property.DeclaringType is IConventionEntityType && property.ClrType == typeof(ObjectId))
+        {
+            var primaryKey = property.FindContainingPrimaryKey();
```
Comment (sanych-sun) [inline src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs:77]: It looks like we found the property as a part of PK in the line #60 and now we "searching" back for PK? Should we add extra parameter and pass PK instead?

Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs:77]: Would have liked to have done that but unfortunately the method signatures are set by EF and not us :(

File: src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs (hunk 3 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs:1]: These files were created this year so put this years date.

I don't think we can put 2010 as you can't backdate copyright and this project was only started in 2023.

Comment (sanych-sun) [reply src/MongoDB.EntityFrameworkCore/Metadata/Conventions/MongoValueGenerationConvention.cs:1]: In Driver we use 2010 everywhere (the year when driver project was started). So should we use 2023 for EF in the same way?

**Pull request merged**